File: m4d/ClientApp/src/pages/searches/SearchesPageModel.ts
import { jsonArrayMember, jsonMember, jsonObject } from "typedjson";

@jsonObject
export class SearchSummary {
  @jsonMember(Number) public id!: number;
  @jsonMember(String) public userName!: string;
  @jsonMember(String) public query?: string;
  @jsonMember(String) public description?: string;
  @jsonMember(String) public searchUrl!: string;
  @jsonMember(String) public searchPageUrl?: string;
  @jsonMember(Number) public mostRecentPage?: number;
  @jsonMember(Number) public count!: number;
  @jsonMember(String) public created!: string;
  @jsonMember(String) public modified!: string;
  @jsonMember(String) public spotify?: string;
  @jsonMember(String) public deleteUrl!: string;
}

@jsonObject
export class SearchesPageModel {
  @jsonArrayMember(SearchSummary) public searches!: SearchSummary[];
  @jsonMember(Number) public page!: number;
  @jsonMember(Number) public totalPages!: number;
  @jsonMember(String) public sort?: string;
  @jsonMember(Boolean) public showDetails!: boolean;
  @jsonMember(Boolean) public spotifyOnly!: boolean;
  @jsonMember(String) public user?: string;
  @jsonMember(Boolean) public isAdmin!: boolean;
  @jsonMember(Boolean) public canDeleteAll!: boolean;
  @jsonMember(String) public basicSearchUrl!: string;
  @jsonMember(String) public advancedSearchUrl!: string;
  @jsonMember(String) public deleteAllUrl?: string;
}

File: m4d/ViewModels/SearchesPageModel.cs
namespace m4d.ViewModels;

/// <summary>
/// Safe-to-serialize projection of a single Search entry for the admin searches index page.
/// URLs are pre-built server-side since they require SongFilter serialization.
/// </summary>
public class SearchSummary
{
    public long Id { get; set; }
    public string UserName { get; set; }
    public string Query { get; set; }
    public string Description { get; set; }
    public string SearchUrl { get; set; }
    public string SearchPageUrl { get; set; }
    public int? MostRecentPage { get; set; }
    public int Count { get; set; }
    public DateTime Created { get; set; }
    public DateTime Modified { get; set; }
    public string Spotify { get; set; }
    public string DeleteUrl { get; set; }
}

/// <summary>
/// Top-level model for the searches Vue page.
/// Server handles paging; Vue renders the current page and builds navigation URLs.
/// </summary>
public class SearchesPageModel
{
    public List<SearchSummary> Searches { get; set; }
    public int Page { get; set; }
    public int TotalPages { get; set; }
    public string Sort { get; set; }
    public bool ShowDetails { get; set; }
    public bool SpotifyOnly { get; set; }
    public string User { get; set; }
    public bool IsAdmin { get; set; }
    public bool CanDeleteAll { get; set; }
    public string BasicSearchUrl { get; set; }
    public string AdvancedSearchUrl { get; set; }
    public string DeleteAllUrl { get; set; }
}
